remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataCacheTest.java

```diff
@@ -0,0 +1,149 @@
+        cache.updateRemoteLogSegmentMetadata(seg2Update);
+        RemoteLogSegmentMetadata segDeleteStarted = seg2.createRemoteLogSegmentWithUpdates(seg2Update);
+
+        // Search should return the above segment.
+        Optional<RemoteLogSegmentMetadata> seg2S250 = cache.remoteLogSegmentMetadata(0, 250);
+        Assertions.assertEquals(seg2.createRemoteLogSegmentWithUpdates(seg2Update), seg2S250.orElse(null));
+
+        // cache.listRemoteLogSegments(0) should contain the above segment.
+        checkContainsAll(cache.listRemoteLogSegments(0), Arrays.asList(segCopyFinished, segDeleteStarted));
+        // But cache.listRemoteLogSegments() should contain all the segments.
+        checkContainsAll(cache.listAllRemoteLogSegments(),
+                Arrays.asList(segCopyInProgress, segCopyFinished, segDeleteStarted));
+
+        // =============================================================================================================
+        // 4.Create a segment and move it to state DELETE_SEGMENT_FINISHED, and check for searching that segment and
+        // listing the segments.
+        // ==============================================================================================================
+        Map<Integer, Long> seg3leaderEpochs = Collections.singletonMap(0, 301L);
+        RemoteLogSegmentId seg3Id = new RemoteLogSegmentId(TP0, Uuid.randomUuid());
+        RemoteLogSegmentMetadata seg3 = new RemoteLogSegmentMetadata(seg3Id, 301L, 400L, -1L, BROKER_ID,
+                System.currentTimeMillis(), SEG_SIZE, seg3leaderEpochs);
+        cache.addToInProgress(seg3);
+        RemoteLogSegmentMetadataUpdate seg3Update1 = new RemoteLogSegmentMetadataUpdate(seg3Id,
+                System.currentTimeMillis(), RemoteLogSegmentState.DELETE_SEGMENT_STARTED, BROKER_ID);
+        cache.updateRemoteLogSegmentMetadata(seg3Update1);
+
+        // Search should return the above segment.
+        Optional<RemoteLogSegmentMetadata> seg3S350 = cache.remoteLogSegmentMetadata(0, 350);
+        Assertions.assertEquals(seg3.createRemoteLogSegmentWithUpdates(seg3Update1), seg3S350.orElse(null));
+
+        RemoteLogSegmentMetadataUpdate seg3Update2 = new RemoteLogSegmentMetadataUpdate(seg3Id,
+                System.currentTimeMillis(), RemoteLogSegmentState.DELETE_SEGMENT_FINISHED, BROKER_ID);
+        cache.updateRemoteLogSegmentMetadata(seg3Update2);
+
+        // cache.listRemoteLogSegments(0) should not contain the above segment.
+        checkContainsAll(cache.listRemoteLogSegments(0), Arrays.asList(segCopyFinished, segDeleteStarted));
+        // But cache.listRemoteLogSegments() should not contain both the segments as it should have been removed.
+        checkContainsAll(cache.listAllRemoteLogSegments(),
+                Arrays.asList(segCopyInProgress, segCopyFinished, segDeleteStarted));
+    }
+
+    private void checkContainsAll(Iterator<RemoteLogSegmentMetadata> allSegments,
```

Comment:
The implementation compromises on the ordering, since it converts the iterator to a set. Is that intentional?

Reply:
Good catch. This is addressed in the latest commit.